src/zkp/piaffg.rs

```diff
@@ -291,50 +271,95 @@ impl Proof for PiAffgProof {
     }
 }
 
+impl PiAffgProof {
+    #[allow(clippy::too_many_arguments)]
+    fn fill_out_transcript(
+        transcript: &mut Transcript,
+        context: &impl ProofContext,
+        input: &PiAffgInput,
+        S: &Commitment,
+        T: &Commitment,
+        A: &Ciphertext,
+        B_x: &CurvePoint,
+        B_y: &Ciphertext,
+        E: &Commitment,
+        F: &Commitment,
+    ) {
+        // First, do Fiat-Shamir consistency check
+        transcript.append_message(b"PiAffg ProofContext", context.as_bytes());
+        transcript.append_message(b"PiAffg CommonInput", &serialize!(&input).unwrap());
+        transcript.append_message(
+            b"(S, T, A, B_x, B_y, E, F)",
+            &[
+                S.to_bytes(),
+                T.to_bytes(),
+                A.to_bytes(),
+                serialize!(&B_x).unwrap(),
+                B_y.to_bytes(),
+                E.to_bytes(),
+                F.to_bytes(),
+            ]
+            .concat(),
+        );
+    }
+}
 #[cfg(test)]
 mod tests {
     use super::*;
     use crate::{
         paillier::DecryptionKey,
         utils::{random_plusminus_by_size_with_minimum, testing::init_testing},
+        zkp::BadContext,
     };
 
     fn random_paillier_affg_proof<R: RngCore + CryptoRng>(
         rng: &mut R,
         x: &BigNumber,
         y: &BigNumber,
-    ) -> Result<()> {
-        let (decryption_key_0, _, _) = DecryptionKey::new(rng)?;
+    ) -> (PiAffgProof, PiAffgInput, Transcript) {
```

Comment:
suggestion: This can still return a `Result<(PiAffgProof, PiAf....)>` so we can use `?` instead of `unwrap`. To me, that's easier to read.